File: sampler.jl
module sampler
export hmc

using LinearAlgebra, Distributions, Statistics, Printf

const RATIOENERGY = 0.1
const RATIODT     = 0.1
const HIGHLEVEL   = 0.9
const LOWLEVEL    = 0.1
const INTERVAL    = 1001
const STEPS       = 3

function getW(q, r, Uqq)
    H = Uqq(q)
    F = eigen(H)
    ve = F.vectors
    e  = F.values
    # L_i = sign(e_i) * |e_i|^{-r}  (0 if e_i==0)
    L = [ei ≈ 0.0 ? 0.0 : sign(ei) * abs(ei)^(-r) for ei in e]
    return ve * Diagonal(L) * ve'
end

function K(p, q, r, Uqq)
    W = getW(q, r, Uqq)
    return 0.5 * dot(p, W * p)
end

function Kp(p, q, r, Uqq)
    W = getW(q, r, Uqq)
    return W * p
end

"""
Hamiltonian Monte Carlo sampler.
Returns (PARTICLES*(ITERATIONS-BURNIN)) × Dim array of samples.
"""
function hmc(U, Uq, Uqq, Dim, BURNIN, ITERATIONS, PARTICLES, r, outbnd, qinit)
    #r = 0.
    # initialise positions
    if !isnothing(qinit)
        qAll = copy(qinit)
    else
        qAll = rand(PARTICLES, Dim)
    end

    # initial total potential energy
    Utotal = sum(U(qAll[i,:]) for i in 1:PARTICLES)
    dt = 1e-9
    Htotal = Utotal ≈ 0.0 ? 1.0 : 2.0 * Utotal

    # storage for samples
    QS = zeros(PARTICLES * (ITERATIONS - BURNIN), Dim)

    for j in 0:(ITERATIONS-1)
        # --- momentum refresh -----------------------------------------
        pAll = rand(Normal(), PARTICLES, Dim)
        KtotalNew = sum(K(pAll[i,:], qAll[i,:], r, Uqq) for i in 1:PARTICLES)
        Utotal    = sum(U(qAll[i,:]) for i in 1:PARTICLES)
        Ktotal    = Htotal - Utotal
        # rescale momenta
        pAll .= pAll .* sqrt(abs(Ktotal / KtotalNew))

        AS = Float64[]    # acceptance probabilities
        ES = []           # energy trajectories (burn-in only)
        anybad = false

        # --- leapfrog / simplified dynamics ---------------------------
        for i in 1:PARTICLES
            bad = false
            p0 = pAll[i,:]
            q0 = qAll[i,:]
            q  = copy(q0)
            p  = copy(p0)

            # burn-in: full leapfrog trajectory
            if j < BURNIN
                UE = Float64[U(q)]
                for step in 1:STEPS
                    bad && break
                    dq = Kp(p, q, r, Uqq)
                    dp = -Uq(q)
                    q .= q .+ dt .* dq
                    p .= p .+ dt .* dp
                    if outbnd(q)
                        bad = true
                        q .= q0
                    else
                        push!(UE, U(q))
                    end
                end
                push!(ES, UE)
                anybad |= bad
            end

            # --- single drift step (always performed) -----------------
            q .= q0
            p .= p0
            dq = Kp(p, q, r, Uqq)
            q .= q .+ sqrt(STEPS) .* dt .* dq
            if outbnd(q)
                bad = true
                q .= q0
            end

            # Metropolis accept/reject
            α = bad ? 0.0 : exp(clamp(U(q0) - U(q), -30.0, 0.0))
            if α < rand()   # always reject if α==0
                q .= q0
            end
            qAll[i,:] .= q
            push!(AS, α)

            # store samples after burn-in
            if j >= BURNIN
                QS[(j - BURNIN)*PARTICLES + i, :] .= q
            end
        end

        # --- burn-in adaptations --------------------------------------
        if j < BURNIN && !anybad
            # unique argmin/argmax along trajectories
            s = unique(argmin.(ES))
            S = unique(argmax.(ES))
            if s == [1, STEPS+1] && S == [1, STEPS+1]
                dt *= (1 + RATIODT)
            elseif s == [1] && S == [STEPS+1]
                dt /= (1 + RATIODT)
            end

            αbar = mean(AS)
            if αbar > HIGHLEVEL
                Htotal = (Htotal - Utotal)*(1 + RATIOENERGY) + Utotal
            elseif αbar < LOWLEVEL
                Htotal = (Htotal - Utotal)/(1 + RATIOENERGY) + Utotal
            end

            if j % INTERVAL == 0
                @printf "%d %.6g %.6g %.6g %.6g %.4g %.4g %s %s\n" j Utotal Ktotal Htotal dt αbar std(AS) s S
            end
        else
            if j % INTERVAL == 0
                @printf "%d %.6g %.6g %.6g %.6g\n" j Utotal Ktotal Htotal dt
            end
        end
    end

    return QS
end

end

File: sampler.py
import numpy as np
from scipy.stats import norm, uniform

RATIOENERGY=0.1
RATIODT=0.1
HIGHLEVEL=0.9
LOWLEVEL=0.1
INTERVAL=1001
STEPS=3

def K(p, q, r, Uqq):
    W = getW(q, r, Uqq)
    return 0.5 * np.dot(p, np.dot(W, p))

def Kp(p, q, r, Uqq):
    W = getW(q, r, Uqq)
    return np.dot(W, p)

def getW(q, r, Uqq):
    eig = np.linalg.eig(Uqq(q))
    ve = eig[1]
    e = eig[0]
    L = np.array([np.sign(e[i]) * np.abs(e[i])**(-r) if e[i] !=0 else 0 for i in range(len(e))])
    return np.dot(np.dot(ve, np.diag(L)), ve.T)

def hmc(U, Uq, Uqq, Dim, BURNIN, ITERATIONS, PARTICLES, outbnd=lambda q:False, qinit=None):
    r = 0.5
    if qinit is not None:
        qAll = qinit
    else:
        #qAll = norm.rvs(0, 1, size=(PARTICLES, Dim))
        qAll = np.random.uniform(0, 1, (PARTICLES, Dim))
    Utotal = np.sum([U(qAll[i]) for i in range(PARTICLES)])
    #dt = np.min([1 / np.sqrt(np.abs(np.linalg.eigvals(Uqq(qAll[i])))) for i in range(PARTICLES)])
    dt =1e-9
    Htotal = 2 * Utotal if Utotal != 0 else 1

    QS = np.zeros((PARTICLES * (ITERATIONS - BURNIN), Dim))

    for j in range(ITERATIONS):
        pAll = norm.rvs(0, 1, size=(PARTICLES, Dim))
        KtotalNew = np.sum([K(pAll[i], qAll[i], r, Uqq) for i in range(PARTICLES)])
        Utotal = np.sum([U(qAll[i]) for i in range(PARTICLES)])
        Ktotal = Htotal - Utotal
        pAll = pAll * np.sqrt(np.abs(Ktotal / KtotalNew))

        AS = []
        ES = []
        anybad = False

        for i in range(PARTICLES):
            bad = False
            p0 = pAll[i]
            q0 = qAll[i]
            q = q0.copy()
            p = q0.copy()
            if j < BURNIN:
                UE = [U(q)]
                for step in range(1, STEPS + 1):
                    if bad:
                        break
                    dq = Kp(p, q, r, Uqq)
                    dp = -Uq(q)
                    q = q + dt * dq
                    p = p + dt * dp
                    if outbnd(q):
                        bad = True
                        q = q0.copy()
                    else:
                        UE.append(U(q))
                ES.append(UE)
                anybad |= bad

            q=q0.copy()
            p=p0.copy()
            dq = Kp(p, q, r, Uqq)
            if False:
                q1 = q + np.sqrt(STEPS) * dt * dq
                dq1 = Kp(p, q1, r, Uqq)
                q = q + np.sqrt(STEPS) * dt * 0.5 * (dq + dq1)
            else:
                q = q + np.sqrt(STEPS) * dt * dq
            if outbnd(q):
                bad = True
                q = q0.copy()

            alpha = 0 if bad else np.exp(np.clip(U(q0) - U(q), -30, 0))
            if alpha < uniform.rvs():
                q = q0
            qAll[i] = q.copy()
            AS.append(alpha)
            if j >= BURNIN:
                QS[(j - BURNIN) * PARTICLES + i] = q.copy()

        if j < BURNIN and not anybad:
            s = np.unique([np.argmin(ES[i]) for i in range(PARTICLES)])
            S = np.unique([np.argmax(ES[i]) for i in range(PARTICLES)])
            if np.array_equal(s, [0, STEPS])and np.array_equal(S, [0, STEPS]):
                dt *= (1 + RATIODT)
            if np.array_equal(s, [0]) and np.array_equal(S, [STEPS]):
                dt /= (1 + RATIODT)

            hi = np.mean(AS) > HIGHLEVEL
            lo = np.mean(AS) < LOWLEVEL
            if hi:
                Htotal = (Htotal - Utotal) * (1 + RATIOENERGY) + Utotal
            if lo:
                Htotal = (Htotal - Utotal) / (1 + RATIOENERGY) + Utotal

            if j % INTERVAL == 0:
                print(f"{j} {Utotal} {Ktotal} {Htotal} {dt} {np.mean(AS)} {np.std(AS)} {s} {S}")
        else:
            if j % INTERVAL == 0:
                print(f"{j} {Utotal} {Ktotal} {Htotal} {dt}")

    return QS
